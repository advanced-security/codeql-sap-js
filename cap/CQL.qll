private import javascript
private import DataFlow

/* TODO: Query bases */
/**
 * Holds if a `DotExpr` ultimately accesses a `SELECT` variable, e.g.
 * ```js
 * SELECT.from
 * SELECT.one.from
 * SELECT.distinct.from
 * ```
 */
private predicate accessesSelect(DotExpr dot) {
  dot.accesses(any(VarRef var | var.getName() = "SELECT"),
    [
      "one", "distinct", "columns", "from", "alias", "where", "having", "groupBy", "orderBy",
      "limit", "forUpdate", "forShareLock"
    ])
  or
  dot.getPropertyName() =
    [
      "one", "distinct", "columns", "from", "alias", "where", "having", "groupBy", "orderBy",
      "limit", "forUpdate", "forShareLock"
    ] and
  accessesSelect(dot.getAChildExpr())
}

/**
 * Method call `SELECT` CQL query expressions, e.g.
 * ```js
 * SELECT.from(Table)
 * SELECT.distinct.from(Table);
 * SELECT.from`Table`.where("col1='*'");
 * SELECT.from(Table).having("col1='*'");
 * ```
 */
private predicate isMethodCallSelect(MethodCallExpr callExpr) {
  callExpr.getCalleeName() =
    [
      "columns", "from", "alias", "where", "having", "groupBy", "orderBy", "limit", "forUpdate",
      "forShareLock"
    ] and
  exists(Expr receiver | receiver = callExpr.getCallee() |
    /*
     * Only property accesses are left up to SELECT, e.g.
     * SELECT.x.y. ...z(cond)
     */

    accessesSelect(receiver)
    or
    /*
     * The immediate prefix is a TaggedTemplateExpr:
     * SELECT.x. ... .z`cond1`.w(cond2)
     */

    exists(TaggedTemplateExpr nestedTaggingExpr |
      receiver.(DotExpr).accesses(nestedTaggingExpr, _)
    |
      isTaggedTemplateSelect(nestedTaggingExpr)
    )
    or
    /*
     * The immediate prefix is a MethodCallExpr:
     * SELECT.x. ... .z(cond1).w(cond2)
     */

    exists(MethodCallExpr nestedCallExpr | receiver.(DotExpr).accesses(nestedCallExpr, _) |
      isMethodCallSelect(nestedCallExpr)
    )
  )
}

/**
 * Tagged `SELECT` CQL query expressions, e.g.
 * ```js
 * SELECT.from`Table`
 * SELECT.distinct.from`Table`;
 * SELECT.from(Table).where`"col1='*'"`;
 * SELECT.from`Table`.having`"col1='*'"`;
 * ```
 */
private predicate isTaggedTemplateSelect(TaggedTemplateExpr tagExpr) {
  exists(Expr taggingExpr |
    taggingExpr = tagExpr.getTag() and
    taggingExpr.(DotExpr).getPropertyName() =
      [
        "columns", "from", "alias", "where", "having", "groupBy", "orderBy", "limit", "forUpdate",
        "forShareLock"
      ]
  |
    /*
     * Only property accesses are left up to SELECT, e.g.
     * SELECT.x.y. ...z`cond`
     */

    accessesSelect(taggingExpr)
    or
    /*
     * The immediate prefix is a TaggedTemplateExpr:
     * SELECT.x. ... .z`cond1`.w`cond2`
     */

    exists(TaggedTemplateExpr nestedTaggingExpr |
      taggingExpr.(DotExpr).accesses(nestedTaggingExpr, _)
    |
      isTaggedTemplateSelect(nestedTaggingExpr)
    )
    or
    /*
     * The immediate prefix is a MethodCallExpr:
     * SELECT.x. ... .z(cond1).w`cond2`
     */

    exists(MethodCallExpr nestedCallExpr | taggingExpr.(DotExpr).accesses(nestedCallExpr, _) |
      isMethodCallSelect(nestedCallExpr)
    )
  )
}

private predicate accessesInsert(DotExpr dot) {
  dot.accesses(any(VarRef var | var.getName() = "INSERT"),
    ["into", "entries", "values", "rows", "as"])
  or
  dot.getPropertyName() = ["into", "entries", "values", "rows", "as"] and
  accessesInsert(dot.getAChildExpr())
}

private predicate isMethodCallInsert(MethodCallExpr callExpr) {
  callExpr.getCalleeName() = ["into", "entries", "values", "rows", "as"] and
  exists(Expr receiver | receiver = callExpr.getCallee() |
    accessesInsert(receiver)
    or
    exists(TaggedTemplateExpr nestedTaggingExpr |
      receiver.(DotExpr).accesses(nestedTaggingExpr, _)
    |
      isTaggedTemplateInsert(nestedTaggingExpr)
    )
    or
    exists(MethodCallExpr nestedCallExpr | receiver.(DotExpr).accesses(nestedCallExpr, _) |
      isMethodCallInsert(nestedCallExpr)
    )
  )
}

private predicate isTaggedTemplateInsert(TaggedTemplateExpr tagExpr) {
  exists(Expr taggingExpr |
    taggingExpr = tagExpr.getTag() and
    taggingExpr.(DotExpr).getPropertyName() = ["into", "entries", "values", "rows", "as"]
  |
    accessesInsert(taggingExpr)
    or
    exists(TaggedTemplateExpr nestedTaggingExpr |
      taggingExpr.(DotExpr).accesses(nestedTaggingExpr, _)
    |
      isTaggedTemplateInsert(nestedTaggingExpr)
    )
    or
    exists(MethodCallExpr nestedCallExpr | taggingExpr.(DotExpr).accesses(nestedCallExpr, _) |
      isMethodCallInsert(nestedCallExpr)
    )
  )
}

newtype TCqlExpr =
  TaggedTemplate(TaggedTemplateExpr tagExpr) or
  MethodCall(MethodCallExpr callExpr)

class CqlExpr extends TCqlExpr {
  TaggedTemplateExpr asTaggedTemplate() { this = TaggedTemplate(result) }

  MethodCallExpr asMethodCall() { this = MethodCall(result) }

  string toString() {
    result = this.asTaggedTemplate().toString() or
    result = this.asMethodCall().toString()
  }

  Location getLocation() {
    result = this.asTaggedTemplate().getLocation() or
    result = this.asMethodCall().getLocation()
  }
}

class CqlSelectExpr extends CqlExpr {
  CqlSelectExpr() {
    isMethodCallSelect(this.asMethodCall()) or isTaggedTemplateSelect(this.asTaggedTemplate())
  }

  predicate selectWhere() {
    this.asMethodCall().getMethodName() = "where" or
    this.asTaggedTemplate().getTag().(DotExpr).getPropertyName() = "where"
  }

  predicate selectFrom() {
    this.asMethodCall().getMethodName() = "from" or
    this.asTaggedTemplate().getTag().(DotExpr).getPropertyName() = "from"
  }
}

class CqlInsertExpr extends CqlExpr {
  CqlInsertExpr() { isMethodCallInsert(this.asMethodCall()) }
}

// class CqlDeleteExpr extends CqlExpr {
//   CqlDeleteExpr() { any() }
// }

// class CqlUpdateExpr extends CqlExpr {
//   CqlUpdateExpr() { any() }
// }

// class CqlUpsertExpr extends CqlExpr {
//   CqlUpsertExpr() { any() }
// }
