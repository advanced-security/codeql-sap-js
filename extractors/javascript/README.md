# Extension to the JavaScript extractor to support CDS compilation

This directory contains a `pre-finalize.sh` script that can be dropped into the `tools` directory
of a JavaScript extractor to enable the automatic compilation and extraction of SAP CAP CDS files
during the `codeql database create` process.

The script requires that the `cds` extractor is available when `codeql database init` is called. This can either be through providing the extractors parent directory using the `--search-path` bundle, or by including the extractor (and this pre-finalize script) in a custom bundle created by the [CodeQL Development Toolkit (qlt)](https://github.com/advanced-security/codeql-development-toolkit).