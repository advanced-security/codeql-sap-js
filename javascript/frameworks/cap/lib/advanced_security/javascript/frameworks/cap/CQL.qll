import javascript
import DataFlow
import advanced_security.javascript.frameworks.cap.CDS

/**
 * Objects from the SQL-like fluent API
 * this is the set of clauses that acts as the base of a statement
 */
class CqlQueryBase extends VarRef {
  CqlQueryBase() {
    exists(string name |
      this.getName() = name and
      name in ["SELECT", "INSERT", "DELETE", "UPDATE", "UPSERT"] and
      (
        /* Made available as a global variable */
        exists(GlobalVariable queryBase | this = queryBase.getAReference())
        or
        /* Imported from `cds.ql` */
        exists(CdsFacade cds |
          cds.getMember("ql").getMember(name).getAValueReachableFromSource().asExpr() = this
        )
      )
    )
  }
}

class CqlSelectBase extends CqlQueryBase {
  CqlSelectBase() { this.getName() = "SELECT" }
}

class CqlInsertBase extends CqlQueryBase {
  CqlInsertBase() { this.getName() = "INSERT" }
}

class CqlDeleteBase extends CqlQueryBase {
  CqlDeleteBase() { this.getName() = "DELETE" }
}

class CqlUpdateBase extends CqlQueryBase {
  CqlUpdateBase() { this.getName() = "UPDATE" }
}

class CqlUpsertBase extends CqlQueryBase {
  CqlUpsertBase() { this.getName() = "UPSERT" }
}

abstract class CqlQueryBaseCall extends CallExpr {
  // TODO: Express "It's a global function or a local function imported from cds.ql"
}

class CqlSelectBaseCall extends CqlQueryBaseCall {
  CqlSelectBaseCall() { this.getCalleeName() = "SELECT" }
}

class CqlInsertBaseCall extends CqlQueryBaseCall {
  CqlInsertBaseCall() { this.getCalleeName() = "INSERT" }
}

/* The cds-ql docs do not mention DELETE being a function acting as a shortcut to any underlying clause. */
class CqlUpdateBaseCall extends CqlQueryBaseCall {
  CqlUpdateBaseCall() { this.getCalleeName() = "UPDATE" }
}

class CqlUpsertBaseCall extends CqlQueryBaseCall {
  CqlUpsertBaseCall() { this.getCalleeName() = "UPSERT" }
}

/**
 * Obtains the receiver across a variety of types of accesses
 */
Expr getRootReceiver(Expr e) {
  result = e and
  (
    e instanceof VarRef
    or
    e instanceof CallExpr and not exists(e.(CallExpr).getReceiver())
  )
  or
  result = getRootReceiver(e.(DotExpr).getBase())
  or
  result = getRootReceiver(e.(MethodCallExpr).getReceiver())
  or
  result = getRootReceiver(e.(PropAccess).getBase())
  or
  result = getRootReceiver(e.(TaggedTemplateExpr).getTag())
}

/**
 * An aggregation type for the two ways to access the fluent API
 * provided by the module cds.ql
 */
private newtype TCqlClause =
  TTaggedTemplate(TaggedTemplateExpr taggedTemplateExpr) {
    exists(CqlQueryBase base | base = getRootReceiver(taggedTemplateExpr)) or
    exists(CqlQueryBaseCall call | call = getRootReceiver(taggedTemplateExpr))
  } or
  TMethodCall(MethodCallExpr callExpr) {
    exists(CqlQueryBase base | base = getRootReceiver(callExpr)) or
    exists(CqlQueryBaseCall call | call = getRootReceiver(callExpr))
  } or
  TShortcutCall(CqlQueryBaseCall callExpr)

class CqlClause extends TCqlClause {
  TaggedTemplateExpr asTaggedTemplate() { this = TTaggedTemplate(result) }

  MethodCallExpr asMethodCall() { this = TMethodCall(result) }

  CallExpr asShortcutCall() { this = TShortcutCall(result) }

  Node flow() { result = this.asExpr().flow() }

  Expr asExpr() {
    result = this.asTaggedTemplate()
    or
    result = this.asMethodCall()
    or
    result = this.asShortcutCall()
  }

  predicate isSelect() {
    /*this.isFinal() and*/
    this.getTypeString() = "SELECT"
  }

  predicate isInsert() { /*this.isFinal() and*/ this.getTypeString() = "INSERT" }

  predicate isUpdate() { /*this.isFinal() and*/ this.getTypeString() = "UPDATE" }

  predicate isUpsert() { /*this.isFinal() and*/ this.getTypeString() = "UPSERT" }

  predicate isDelete() { /*this.isFinal() and*/ this.getTypeString() = "DELETE" }

  predicate isRead() { this.isSelect() }

  predicate isWrite() { this.isInsert() or this.isUpdate() or this.isUpsert() or this.isDelete() }

  Expr getArgument() {
    result = this.asTaggedTemplate().getTemplate() or
    result = this.asMethodCall().getAnArgument() or
    result = this.asShortcutCall().getAnArgument()
  }

  /**
   * Convert this `CqlClause` into a `DotExpr`, i.e.
   * get `` SELECT.from`Table`.where `` when given `` SELECT.from`Table`.where`cond` ``.
   */
  DotExpr asDotExpr() {
    result = this.asTaggedTemplate().getTag() or
    result = this.asMethodCall().getCallee()
  }

  string toString() {
    result = this.asTaggedTemplate().toString() or
    result = this.asMethodCall().toString() or
    result = this.asShortcutCall().toString()
  }

  string getTypeString() {
    result = getRootReceiver(this.asTaggedTemplate()).(VarRef).getName() or
    result = getRootReceiver(this.asTaggedTemplate()).(CallExpr).getCalleeName() or
    result = getRootReceiver(this.asMethodCall()).(VarRef).getName() or
    result = getRootReceiver(this.asMethodCall()).(CallExpr).getCalleeName()
  }

  Location getLocation() {
    result = this.asTaggedTemplate().getLocation() or
    result = this.asMethodCall().getLocation() or
    result = this.asShortcutCall().getLocation()
  }

  CqlQueryBase getCqlBase() { result = getRootReceiver(this.asMethodCall()) }

  CqlQueryBaseCall getCqlBaseCall() {
    result = getRootReceiver(this.asTaggedTemplate()) or
    result = getRootReceiver(this.asMethodCall())
  }

  /** Describes a parent expression relation */
  Expr getParentExpr() {
    result = this.asTaggedTemplate().getParentExpr() or
    result = this.asMethodCall().getParentExpr() or
    result = this.asShortcutCall().getParentExpr()
  }

  /**
   * Possible cases for constructing a chain of clauses:
   *
   * (looking at the terminal clause and its possible parent types as tuples: (this, parent))
   * 1. TMethodCall.TMethodCall
   *     - example `(SELECT.from(Table),  SELECT.from(Table).where("col1='*'"))`
   * 2. TShortcutCall.TMethodCall
   *     - example `(SELECT("col1, col2"), SELECT("col1, col2").from("Table"))`
   *
   * Note that ShortcutCalls cannot be added to any clause chain other than the first position, e.g.
   * ``` javascript
   * SELECT("col1, col2").INSERT(col2)  // Invalid!
   * ```
   */
  CqlClause getParentCqlClause() {
    /* ========== The parent is a shortcut call ========== */
    result.asShortcutCall() = this.asTaggedTemplate().getParentExpr().getParentExpr() or
    result.asShortcutCall() = this.asMethodCall().getParentExpr().getParentExpr() or
    result.asShortcutCall() = this.asShortcutCall().getParentExpr().getParentExpr() or
    /* ========== The parent is a tagged template ========== */
    result.asTaggedTemplate() = this.asTaggedTemplate().getParentExpr().getParentExpr() or
    result.asTaggedTemplate() = this.asMethodCall().getParentExpr().getParentExpr() or
    result.asTaggedTemplate() = this.asShortcutCall().getParentExpr().getParentExpr() or
    /* ========== The parent is a method call ========== */
    result.asMethodCall() = this.asTaggedTemplate().getParentExpr().getParentExpr() or
    result.asMethodCall() = this.asMethodCall().getParentExpr().getParentExpr() or
    result.asMethodCall() = this.asShortcutCall().getParentExpr().getParentExpr()
  }

  Expr getAnAncestorExpr() {
    result = this.asTaggedTemplate().getParentExpr+() or
    result = this.asMethodCall().getParentExpr+() or
    result = this.asShortcutCall().getParentExpr+()
  }

  CqlClause getAnAncestorCqlClause() {
    result.asTaggedTemplate() = this.getAnAncestorExpr() or
    result.asMethodCall() = this.getAnAncestorExpr() or
    result.asShortcutCall() = this.getAnAncestorExpr()
  }

  /** Describes a child expression relation */
  Expr getAChildExpr() {
    result = this.asTaggedTemplate().getAChildExpr() or
    result = this.asMethodCall().getAChildExpr() or
    result = this.asShortcutCall().getAChildExpr()
  }

  /**
   * the same chain order logic as `getParentCqlClause` but reversed
   */
  CqlClause getAChildCqlClause() {
    /* ========== The parent is a shortcut call ========== */
    result.asShortcutCall() = this.asTaggedTemplate().getAChildExpr().getAChildExpr() or
    result.asShortcutCall() = this.asMethodCall().getAChildExpr().getAChildExpr() or
    result.asShortcutCall() = this.asShortcutCall().getAChildExpr().getAChildExpr() or
    /* ========== The parent is a tagged template ========== */
    result.asTaggedTemplate() = this.asTaggedTemplate().getAChildExpr().getAChildExpr() or
    result.asTaggedTemplate() = this.asMethodCall().getAChildExpr().getAChildExpr() or
    result.asTaggedTemplate() = this.asShortcutCall().getAChildExpr().getAChildExpr() or
    /* ========== The parent is a method call ========== */
    result.asMethodCall() = this.asTaggedTemplate().getAChildExpr().getAChildExpr() or
    result.asMethodCall() = this.asMethodCall().getAChildExpr().getAChildExpr() or
    result.asMethodCall() = this.asShortcutCall().getAChildExpr().getAChildExpr()
  }

  Expr getADescendantExpr() {
    result = this.asTaggedTemplate().getAChildExpr+() or
    result = this.asMethodCall().getAChildExpr+() or
    result = this.asShortcutCall().getAChildExpr+()
  }

  CqlClause getADescendantCqlClause() {
    result.asTaggedTemplate() = this.getADescendantExpr() or
    result.asMethodCall() = this.getADescendantExpr() or
    result.asShortcutCall() = this.getADescendantExpr()
  }

  predicate isFinal() { not exists(this.getParentCqlClause()) }

  /**
   * Matches the given `CqlClause` to its method/property name, nested at arbitrary depth.
   */
  string getAnAPIName() {
    result = this.asDotExpr().getPropertyName() or
    result = this.getADescendantCqlClause().getAnAPIName()
  }

  abstract CqlClause getEntityAccessingClause();

  /**
   * Gets the reference to the entity that this clause is accessing.
   */
  ExprNode getAccessingEntityReference() {
    result = this.getEntityAccessingClause().getArgument().flow()
  }

  /**
   * Gets the reference of the service that runs this CQL clause.
   */
  ServiceInstance getRunner() {
    exists(CdsTransaction tx | this = tx.getAnExecutedCqlClause() and result = tx.getRunner())
    or
    exists(SrvRun srvRun | this = srvRun.getCql() and result = srvRun.getRecipient())
  }

  CdlEntity getAccessingEntityDefinition() {
    /* 1. String literals or template strings */
    result.getName() =
      this.getAccessingEntityReference().(EntityReferenceFromCqlClause).getStringValue()
    or
    result.getUnqualifiedName() =
      this.getAccessingEntityReference().(EntityReferenceFromCqlClause).getStringValue() and
    /* The entity is accessed by its own service. */
    result = this.getRunner().getDefinition().getCdsDeclaration().getAnEntity()
    or
    /* 2. Variable whose value is a reference to an entity */
    result =
      this.getAccessingEntityReference()
          .getALocalSource()
          .(EntityReferenceFromEntities)
          .getCqlDefinition()
  }
}

class CqlSelectClause extends CqlClause {
  CqlSelectClause() { this.isSelect() }

  override CqlSelectClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    result.asDotExpr().getPropertyName() = "from"
  }
}

class CqlInsertClause extends CqlClause {
  CqlInsertClause() { this.isInsert() }

  override CqlInsertClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    result.asDotExpr().getPropertyName() = "into"
  }
}

class CqlUpdateClause extends CqlClause {
  CqlUpdateClause() { this.isUpdate() }

  override CqlUpdateClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    (
      result.asDotExpr().getPropertyName() = "entity" or
      exists(result.asShortcutCall())
    )
  }
}

class CqlUpsertClause extends CqlClause {
  CqlUpsertClause() { this.isUpsert() }

  override CqlUpsertClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    result.asDotExpr().getPropertyName() = "into"
  }
}

class CqlDeleteClause extends CqlClause {
  CqlDeleteClause() { this.isDelete() }

  override CqlDeleteClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    result.asDotExpr().getPropertyName() = "from"
  }
}
