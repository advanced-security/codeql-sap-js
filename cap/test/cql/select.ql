import javascript
import CQL

// select any(CqlSelectExpr s)
select "hihi"