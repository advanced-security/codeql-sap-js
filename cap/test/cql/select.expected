| select.js:2:14:2:26 | SELECT`Table` |
| select.js:5:14:5:35 | SELECT. ... `Table` |
| select.js:7:14:7:40 | SELECT. ... `Table` |
| select.js:12:14:12:31 | SELECT.from`Table` |
| select.js:12:14:12:51 | SELECT. ... , col2` |
| select.js:13:14:13:31 | SELECT.from`Table` |
| select.js:16:14:16:31 | SELECT.from`Table` |
| select.js:16:14:16:68 | SELECT. ... lias }` |
| select.js:17:14:17:31 | SELECT.from`Table` |
| select.js:18:14:18:31 | SELECT.from`Table` |
| select.js:22:14:22:31 | SELECT.from`Table` |
| select.js:43:14:43:31 | SELECT.from`Table` |
| select.js:44:14:44:31 | SELECT.from`Table` |
| select.js:45:14:45:31 | SELECT.from`Table` |
| select.js:45:14:45:49 | SELECT. ... 1='*'"` |
| select.js:46:14:46:31 | SELECT.from`Table` |
| select.js:47:14:47:31 | SELECT.from`Table` |
| select.js:47:14:47:51 | SELECT. ... ${"*"}` |
| select.js:48:14:48:31 | SELECT.from`Table` |
| select.js:49:14:49:31 | SELECT.from`Table` |
| select.js:49:14:49:61 | SELECT. ...  10)]}` |
| select.js:50:14:50:31 | SELECT.from`Table` |
| select.js:53:14:53:31 | SELECT.from`Table` |
| select.js:54:14:54:31 | SELECT.from`Table` |
| select.js:54:14:54:51 | SELECT. ... , col2` |
| select.js:55:14:55:31 | SELECT.from`Table` |
| select.js:56:14:56:31 | SELECT.from`Table` |
| select.js:56:14:56:63 | SELECT. ... .prop2` |
| select.js:57:14:57:31 | SELECT.from`Table` |
| select.js:63:14:63:31 | SELECT.from`Table` |
| select.js:64:14:64:31 | SELECT.from`Table` |
| select.js:65:14:65:31 | SELECT.from`Table` |
| select.js:65:14:65:50 | SELECT. ... 1='*'"` |
| select.js:66:14:66:31 | SELECT.from`Table` |
| select.js:67:14:67:31 | SELECT.from`Table` |
| select.js:67:14:67:52 | SELECT. ... ${"*"}` |
| select.js:68:14:68:31 | SELECT.from`Table` |
| select.js:69:14:69:31 | SELECT.from`Table` |
| select.js:69:14:69:62 | SELECT. ...  10)]}` |
| select.js:70:14:70:31 | SELECT.from`Table` |
| select.js:72:14:72:31 | SELECT.from`Table` |
| select.js:73:14:73:31 | SELECT.from`Table` |
| select.js:74:14:74:31 | SELECT.from`Table` |
| select.js:75:14:75:31 | SELECT.from`Table` |
| select.js:76:14:76:31 | SELECT.from`Table` |
| select.js:77:14:77:31 | SELECT.from`Table` |
| select.js:80:14:80:31 | SELECT.from`Table` |
| select.js:83:14:83:31 | SELECT.from`Table` |
| select.js:87:14:87:31 | SELECT.from`Table` |
| select.js:87:14:87:51 | SELECT. ... , col2` |
| select.js:88:14:88:31 | SELECT.from`Table` |
| select.js:88:14:88:51 | SELECT. ... , col2` |
| select.js:89:14:89:31 | SELECT.from`Table` |
| select.js:89:14:89:51 | SELECT. ... , col2` |
| select.js:89:14:89:70 | SELECT. ... 1='*'"` |
| select.js:90:14:90:31 | SELECT.from`Table` |
| select.js:90:14:90:51 | SELECT. ... , col2` |
| select.js:91:14:91:31 | SELECT.from`Table` |
| select.js:91:14:91:51 | SELECT. ... , col2` |
| select.js:91:14:91:72 | SELECT. ... ${"*"}` |
| select.js:92:14:92:31 | SELECT.from`Table` |
| select.js:92:14:92:51 | SELECT. ... , col2` |
| select.js:93:14:93:31 | SELECT.from`Table` |
| select.js:93:14:93:51 | SELECT. ... , col2` |
| select.js:93:14:95:3 | SELECT. ... 0),\\n]}` |
| select.js:96:14:96:31 | SELECT.from`Table` |
| select.js:96:14:96:51 | SELECT. ... , col2` |
| select.js:101:14:101:31 | SELECT.from`Table` |
| select.js:104:14:104:31 | SELECT.from`Table` |
| select.js:107:14:107:31 | SELECT.from`Table` |
| select.js:109:14:109:31 | SELECT.from`Table` |
| select.js:112:14:112:31 | SELECT.from`Table` |
| select.js:114:14:114:31 | SELECT.from`Table` |
| select.js:117:14:117:31 | SELECT.from`Table` |
| select.js:119:14:119:31 | SELECT.from`Table` |
| select.js:123:14:123:31 | SELECT.from`Table` |
| select.js:123:14:123:63 | SELECT. ... .prop2` |
| select.js:126:14:126:31 | SELECT.from`Table` |
| select.js:126:14:126:63 | SELECT. ... .prop2` |
| select.js:129:14:129:31 | SELECT.from`Table` |
| select.js:129:14:129:63 | SELECT. ... .prop2` |
| select.js:129:14:130:21 | SELECT. ... 1='*'"` |
| select.js:131:14:131:31 | SELECT.from`Table` |
| select.js:131:14:131:63 | SELECT. ... .prop2` |
| select.js:135:14:135:31 | SELECT.from`Table` |
| select.js:135:14:135:63 | SELECT. ... .prop2` |
| select.js:135:14:136:23 | SELECT. ... ${"*"}` |
| select.js:137:14:137:31 | SELECT.from`Table` |
| select.js:137:14:137:63 | SELECT. ... .prop2` |
| select.js:140:14:140:31 | SELECT.from`Table` |
| select.js:140:14:140:63 | SELECT. ... .prop2` |
| select.js:140:14:141:33 | SELECT. ...  10)]}` |
| select.js:142:14:142:31 | SELECT.from`Table` |
| select.js:142:14:142:63 | SELECT. ... .prop2` |
| select.js:147:14:147:31 | SELECT.from`Table` |
| select.js:150:14:150:31 | SELECT.from`Table` |
| select.js:153:14:153:31 | SELECT.from`Table` |
| select.js:157:14:157:31 | SELECT.from`Table` |
| select.js:160:14:160:31 | SELECT.from`Table` |
| select.js:164:14:164:31 | SELECT.from`Table` |
| select.js:167:14:167:31 | SELECT.from`Table` |
| select.js:171:14:171:31 | SELECT.from`Table` |
| select.js:176:14:176:31 | SELECT.from`Table` |
| select.js:176:14:176:63 | SELECT. ... .prop2` |
| select.js:177:14:177:31 | SELECT.from`Table` |
| select.js:177:14:177:61 | SELECT. ... .prop2` |
| select.js:178:14:178:31 | SELECT.from`Table` |
| select.js:178:14:178:61 | SELECT. ... l2 asc` |
| select.js:179:14:179:31 | SELECT.from`Table` |
| select.js:179:14:179:58 | SELECT. ... l2 asc` |
| select.js:180:14:180:31 | SELECT.from`Table` |
| select.js:180:14:180:63 | SELECT. ... .prop2` |
| select.js:181:14:181:31 | SELECT.from`Table` |
| select.js:181:14:181:62 | SELECT. ... .prop2` |
| select.js:182:14:182:31 | SELECT.from`Table` |
| select.js:182:14:182:62 | SELECT. ... 2 desc` |
| select.js:183:14:183:31 | SELECT.from`Table` |
| select.js:183:14:183:60 | SELECT. ... 2 desc` |
| select.js:186:14:186:31 | SELECT.from`Table` |
| select.js:187:14:187:31 | SELECT.from`Table` |
| select.js:188:14:188:31 | SELECT.from`Table` |
| select.js:189:14:189:31 | SELECT.from`Table` |
| select.js:190:14:190:31 | SELECT.from`Table` |
| select.js:191:14:191:31 | SELECT.from`Table` |
| select.js:196:14:196:31 | SELECT.from`Table` |
| select.js:196:14:196:51 | SELECT. ... , col2` |
| select.js:196:14:197:23 | SELECT. ... ${"*"}` |
| select.js:200:14:200:31 | SELECT.from`Table` |
| select.js:200:14:200:51 | SELECT. ... , col2` |
| select.js:200:14:201:23 | SELECT. ... ${"*"}` |
| select.js:204:14:204:40 | SELECT. ... `Table` |
| select.js:204:14:204:70 | SELECT. ...  10)]}` |
